perf/src/deduper.rs

```diff
@@ -17,35 +17,46 @@ pub struct Deduper<const K: usize> {
     bits: Vec<AtomicU64>,
     seeds: [(u128, u128); K],
     clock: Instant,
-    // Maximum number of one bits before the false positive
-    // rate exceeds the specified threshold.
-    capacity: u64,
     popcount: AtomicU64, // Number of one bits in self.bits.
 }
 
 impl<const K: usize> Deduper<K> {
-    pub fn new<R: Rng>(rng: &mut R, false_positive_rate: f64, num_bits: u64) -> Self {
-        assert!(0.0 < false_positive_rate && false_positive_rate < 1.0);
-        let size = usize::try_from(num_bits.checked_add(63).unwrap() / 64).unwrap();
-        let capacity = num_bits as f64 * false_positive_rate.powf(1f64 / K as f64);
+    pub fn new<R: Rng>(rng: &mut R, num_bits: u64) -> Self {
+        let size = num_bits.checked_add(63).unwrap() / 64;
+        let size = usize::try_from(size).unwrap();
         Self {
             num_bits,
             seeds: std::array::from_fn(|_| rng.gen()),
             clock: Instant::now(),
             bits: repeat_with(AtomicU64::default).take(size).collect(),
-            capacity: capacity as u64,
             popcount: AtomicU64::default(),
         }
     }
 
-    pub fn maybe_reset<R: Rng>(&mut self, rng: &mut R, reset_cycle: &Duration) {
+    fn false_positive_rate(&self) -> f64 {
         let popcount = self.popcount.load(Ordering::Relaxed);
-        if popcount >= self.capacity || &self.clock.elapsed() >= reset_cycle {
+        let ones_ratio = popcount.min(self.num_bits) as f64 / self.num_bits as f64;
+        ones_ratio.powi(K as i32)
+    }
+
+    /// Resets the Deduper if either it is older than the reset_cycle or it is
+    /// saturated enough that false positive rate exceeds specified threshold.
+    /// Returns true if the deduper was saturated.
+    pub fn maybe_reset<R: Rng>(
+        &mut self,
+        rng: &mut R,
+        false_positive_rate: f64,
+        reset_cycle: Duration,
+    ) -> bool {
```

Comment:
Given that the function is named `maybe_reset`, I would expect the return value to indicate whether the `Deduper` was reset instead of saturated. This isn't a blocker for me for this PR, I just didn't read the comment very closely on the first pass, and then was wondering how `maybe_reset()` was returning false with `Duration::from_millis(0), // reset_cycle` in the unit test.

It looks like the return value is only being used for unit tests, and for these, we could use `Duration::from_millis(std::u64::MAX)` to isolate behavior between reset cycle elapsed or saturation

Reply:
This is only for the sake of collecting saturation metrics in https://github.com/solana-labs/solana/pull/30779
i.e. we want to track how often the deduper is saturated before its reset cycle.